### ISSUE TITLE
`make -j mana` can fail with `p2p_deterministic.c`, where `make mana` does not.

### ISSUE THREAD

**Issue by leonidbelyaev** [mpickpt/mana#365]
When I build mana using `make -j mana`, I sometimes get:

```
[...]
mpic++ -std=c++14 -fno-stack-protector -Wall -g3 -O0 -DDEBUG -fPIC -I../../dmtcp/include -I../../dmtcp/jalib -I.. -I../../dmtcp/src -I../lower-half -c -o mpi_p2p_wrappers.o mpi_p2p_wrappers.cpp
mpic++ -std=c++14 -fno-stack-protector -Wall -g3 -O0 -DDEBUG -fPIC -I../../dmtcp/include -I../../dmtcp/jalib -I.. -I../../dmtcp/src -I../lower-half -c -o mpi_request_wrappers.o mpi_request_wrappers.cpp
p2p-deterministic.c: In function ‘set_next_msg’:
p2p-deterministic.c:154:34: error: ‘P2P_LOG_MSG’ undeclared (first use in this function)
  154 |     snprintf(buf, sizeof(buf)-1, P2P_LOG_MSG, rank);
      |                                  ^~~~~~~~~~~
p2p-deterministic.c:164:26: error: invalid application of ‘sizeof’ to incomplete type ‘struct p2p_log_msg’
  164 |   p2p_msg = malloc(sizeof(*p2p_msg) + size * count);
      |                          ^
p2p-deterministic.c:165:10: error: invalid use of undefined type ‘struct p2p_log_msg’
  165 |   p2p_msg->count = count;
      |          ^~
p2p-deterministic.c:166:10: error: invalid use of undefined type ‘struct p2p_log_msg’
  166 |   p2p_msg->datatype = datatype;
      |          ^~
p2p-deterministic.c:167:10: error: invalid use of undefined type ‘struct p2p_log_msg’
  167 |   p2p_msg->source = source;
      |          ^~
p2p-deterministic.c:168:10: error: invalid use of undefined type ‘struct p2p_log_msg’
  168 |   p2p_msg->tag = tag;
      |          ^~
p2p-deterministic.c:169:10: error: invalid use of undefined type ‘struct p2p_log_msg’
  169 |   p2p_msg->comm = comm;
      |          ^~
p2p-deterministic.c:170:17: error: invalid use of undefined type ‘struct p2p_log_msg’
  170 |   memcpy(p2p_msg->data, data, size * count);
      |                 ^~
p2p-deterministic.c:173:10: error: invalid use of undefined type ‘struct p2p_log_msg’
  173 |   p2p_msg->request = (request ? *request : MPI_REQUEST_NULL);
      |          ^~
p2p-deterministic.c:175:12: error: invalid use of undefined type ‘struct p2p_log_msg’
  175 |     p2p_msg->source = status->MPI_SOURCE;
      |            ^~
p2p-deterministic.c:176:12: error: invalid use of undefined type ‘struct p2p_log_msg’
  176 |     p2p_msg->tag = status->MPI_TAG;
      |            ^~
p2p-deterministic.c:183:3: warning: implicit declaration of function ‘writeall’ [-Wimplicit-function-declaration]
  183 |   writeall(fd, p2p_msg, sizeof(*p2p_msg) + size * count);
      |   ^~~~~~~~
p2p-deterministic.c:183:31: error: invalid application of ‘sizeof’ to incomplete type ‘struct p2p_log_msg’
  183 |   writeall(fd, p2p_msg, sizeof(*p2p_msg) + size * count);
      |                               ^
p2p-deterministic.c: In function ‘p2p_replay_pre_irecv’:
p2p-deterministic.c:194:22: error: storage size of ‘p2p_msg’ isn’t known
  194 |   struct p2p_log_msg p2p_msg;
      |                      ^~~~~~~
p2p-deterministic.c:194:22: warning: unused variable ‘p2p_msg’ [-Wunused-variable]
p2p-deterministic.c: In function ‘p2p_replay_post_iprobe’:
p2p-deterministic.c:209:24: error: storage size of ‘p2p_msg’ isn’t known
  209 |     struct p2p_log_msg p2p_msg;
      |                        ^~~~~~~
p2p-deterministic.c:209:24: warning: unused variable ‘p2p_msg’ [-Wunused-variable]
p2p-deterministic.c: In function ‘save_request_info’:
p2p-deterministic.c:253:26: error: storage size of ‘p2p_request’ isn’t known
  253 |   struct p2p_log_request p2p_request;
      |                          ^~~~~~~~~~~
p2p-deterministic.c:260:34: error: ‘P2P_LOG_REQUEST’ undeclared (first use in this function)
  260 |     snprintf(buf, sizeof(buf)-1, P2P_LOG_REQUEST, rank);
      |                                  ^~~~~~~~~~~~~~~
p2p-deterministic.c:253:26: warning: unused variable ‘p2p_request’ [-Wunused-variable]
  253 |   struct p2p_log_request p2p_request;
      |                          ^~~~~~~~~~~
p2p-deterministic.c: At top level:
p2p-deterministic.c:38:27: error: storage size of ‘next_msg_entry’ isn’t known
   38 | static struct p2p_log_msg next_msg_entry = {0, MPI_CHAR, 0, 0, 0, MPI_COMM_NULL, MPI_REQUEST_NULL};
      |                           ^~~~~~~~~~~~~~
[...]
```

And the build fails. 

This does not happen if I use `make mana` instead.

**Comment by gc00** [comment]
@JainTwinkle , would you like to add a commit in PR #364 to fix this issue also?

@leonidbelyaev , Really nice catch!

If you look at `mpi-wrappers/Makefile`, the issue seems to be that the .h file has the correct dependency.
> p2p-deterministic.h: p2p-deterministic.txt p2p-deterministic.py

But the `.c.o:` target says that p2p-deterministic.c doesn't depend on any .h file.  So, p2p-deterministic.h hasn't neccessarily been built yet.

The fix is to add to the Makefile something like:
> p2p-deterministic.o: p2p-deterministic.c p2p-deterministic.h
 \<TAB\> ${MPICC} ${CFLAGS} -c -o $@ $<

I haven't tested this.  @JainTwinkle or @leonidbelyaev , please test and confirm this fix.

**Comment by JainTwinkle** [comment]
@gc00 @leonidbelyaev  Thanks, I will add the fix in PR #364 before merging into the main branch.

**Comment by leonidbelyaev** [comment]
@gc00 @JainTwinkle I just tried this fix several times, and did not get this race condition to occur. Seems like it works.
